libbpf-tools/tcppktlat.bpf.c

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: GPL-2.0
+// Copyright (c) 2023 Wenbo Zhang
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+
+#include "compat.bpf.h"
+#include "tcppktlat.h"
+
+#define MAX_ENTRIES	10240
+#define AF_INET		2
+
+const volatile pid_t targ_pid = 0;
+const volatile pid_t targ_tid = 0;
+const volatile __u16 targ_sport = 0;
+const volatile __u16 targ_dport = 0;
+const volatile __u64 targ_min_us = 0;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_ENTRIES);
+	__type(key, u64);
+	__type(value, u64);
+} start SEC(".maps");
+
+SEC("tracepoint/tcp/tcp_probe")
+int tcp_probe_btf(struct trace_event_raw_tcp_probe *ctx)
+{
+	u64 sock_cookie, ts;
+
+	if (targ_sport && targ_sport != BPF_CORE_READ(ctx, sport))
+		return 0;
+	if (targ_dport && targ_dport != BPF_CORE_READ(ctx, dport))
+		return 0;
+	if (BPF_CORE_READ(ctx, data_len) == 0)
+		return 0;
+	sock_cookie = BPF_CORE_READ(ctx, sock_cookie);
+	ts = bpf_ktime_get_ns();
+	bpf_map_update_elem(&start, &sock_cookie, &ts, 0);
+	return 0;
+}
+
+SEC("tracepoint/tcp/tcp_rcv_space_adjust")
+int tcp_rcv_space_adjust_btf(struct trace_event_raw_tcp_event_sk *ctx)
+{
+	u64 id = bpf_get_current_pid_tgid(), *tsp;
+	u32 pid = id >> 32, tid = id;
+	struct event *eventp;
+	u64 sock_cookie;
+	s64 delta_us;
+	u16 family;
+
+	if (targ_pid && targ_pid != pid)
+		return 0;
+	if (targ_tid && targ_tid != tid)
+		return 0;
+
+	sock_cookie = BPF_CORE_READ(ctx, sock_cookie);
+	tsp = bpf_map_lookup_elem(&start, &sock_cookie);
```

Comment:
> So, if data_len is zero, tcp_rcv_space_adjust will not be triggered, right ?

Yes.

> We should do bpf_map_lookup_elem before checking pid/tid.
> 
> And if pid/tid not matched, go to cleanup.

Good catch, thanks!

Reply:
On old kernel, there is no `family` field for tracepoint `tcp_rcv_space_adjust`.
```
name: tcp_rcv_space_adjust
ID: 1378
format:
        field:unsigned short common_type;       offset:0;       size:2; signed:0;
        field:unsigned char common_flags;       offset:2;       size:1; signed:0;
        field:unsigned char common_preempt_count;       offset:3;       size:1; signed:0;
        field:int common_pid;   offset:4;       size:4; signed:1;

        field:const void * skaddr;      offset:8;       size:8; signed:0;
        field:__u16 sport;      offset:16;      size:2; signed:0;
        field:__u16 dport;      offset:18;      size:2; signed:0;
        field:__u8 saddr[4];    offset:20;      size:4; signed:0;
        field:__u8 daddr[4];    offset:24;      size:4; signed:0;
        field:__u8 saddr_v6[16];        offset:28;      size:16;        signed:0;
        field:__u8 daddr_v6[16];        offset:44;      size:16;        signed:0;
        field:__u64 sock_cookie;        offset:64;      size:8; signed:0;

print fmt: "sport=%hu dport=%hu saddr=%pI4 daddr=%pI4 saddrv6=%pI6c daddrv6=%pI6c sock_cookie=%llx", REC->sport, REC->dport, REC->saddr, REC->daddr, REC->saddr_v6, REC->daddr_v6, REC->sock_cookie
```

Reply:
> On old kernel, there is no `family` field for tracepoint `tcp_rcv_space_adjust`.
> 
> ```
> name: tcp_rcv_space_adjust
> ID: 1378
> format:
>         field:unsigned short common_type;       offset:0;       size:2; signed:0;
>         field:unsigned char common_flags;       offset:2;       size:1; signed:0;
>         field:unsigned char common_preempt_count;       offset:3;       size:1; signed:0;
>         field:int common_pid;   offset:4;       size:4; signed:1;
> 
>         field:const void * skaddr;      offset:8;       size:8; signed:0;
>         field:__u16 sport;      offset:16;      size:2; signed:0;
>         field:__u16 dport;      offset:18;      size:2; signed:0;
>         field:__u8 saddr[4];    offset:20;      size:4; signed:0;
>         field:__u8 daddr[4];    offset:24;      size:4; signed:0;
>         field:__u8 saddr_v6[16];        offset:28;      size:16;        signed:0;
>         field:__u8 daddr_v6[16];        offset:44;      size:16;        signed:0;
>         field:__u64 sock_cookie;        offset:64;      size:8; signed:0;
> 
> print fmt: "sport=%hu dport=%hu saddr=%pI4 daddr=%pI4 saddrv6=%pI6c daddrv6=%pI6c sock_cookie=%llx", REC->sport, REC->dport, REC->saddr, REC->daddr, REC->saddr_v6, REC->daddr_v6, REC->sock_cookie
> ```

Get it, I find there's already a BPF helper: [BPF_FUNC_get_socket_cookie()](https://github.com/torvalds/linux/commit/91b8270f2a4d1d9b268de90451cdca63a70052d6) from v4.12. So I think we can change `tracepoint` to `raw_tp` && `tp_btf`  to resolve this.

Reply:
Great! `raw_tp` is what I have in mind, but I didn't aware `bpf_get_socket_cookie()`.

Reply:
@chenhengqi Addressed, PTAL again, thanks!    :)

Reply:
bpf_get_socket_cookie() is not avaliable for tracing progs. :(

Reply:
> bpf_get_socket_cookie() is not avaliable for tracing progs. :(

@chenhengqi Oh, I see. Tracing is only supported since version v5.12. so, any suggestions?